Add test for OS updates being staged within 5-10m.

dgoodwin opened the pull request:
Recent test failure investigation led to a possibility of slow disk
speeds doing OS content extraction. Add a test to ensure that we move
from OSUpdateStarted to OSUpdateStaged within 5-10 minutes, which
primarily just encompasses extraction of the content.

If any node is over 10 minutes we will fail the test, if some are over 5
we'll flake instead.

deads2k commented on pkg/synthetictests/operators.go:
```diff
@@ -73,6 +73,77 @@ func testOperatorStateTransitions(events monitorapi.Intervals, conditionTypes []
+	success := &junitapi.JUnitTestCase{Name: testName}
+
+	type startedStaged struct {
+		OSUpdateStarted time.Time
```
give me some comments about what this state means

deads2k commented on pkg/synthetictests/operators.go:
```diff
@@ -73,6 +73,77 @@ func testOperatorStateTransitions(events monitorapi.Intervals, conditionTypes []
+
+	type startedStaged struct {
+		OSUpdateStarted time.Time
+		OSUpdateStaged  time.Time
```
give me some comments about what this state means

deads2k commented:
/approve

a request for some comments and lgtm

wking commented on pkg/synthetictests/operators.go:
```diff
@@ -73,6 +73,82 @@ func testOperatorStateTransitions(events monitorapi.Intervals, conditionTypes []
+		if ss.OSUpdateStaged.IsZero() {
+			slowStageMessages = append(slowStageMessages, fmt.Sprintf("%s OSUpdateStarted at %s, did not make it to OSUpdateStaged", node, ss.OSUpdateStarted))
+			failTest = true
+		} else if ss.OSUpdateStaged.Sub(ss.OSUpdateStarted) > 5*time.Minute {
```
possible worth an additional case here for `ss.OSUpdateStarted.IsZero()` for "best-effort event submission must have dropped the started event, so I dunno how long this node too to get from there to staged".  I guess I'd expect that to be a flake, but should be rare enough that if we decide to fail on it the number of runs that fail as a result should be small.

dgoodwin commented:
@wking has pointed out that we may have suites that do multi-stage updates, meaning we move through these conditions several times. I'm also not sure this currently properly handles install time events. Will revisit asap tomorrow.

wking commented on pkg/synthetictests/operators.go:
```diff
@@ -73,6 +73,82 @@ func testOperatorStateTransitions(events monitorapi.Intervals, conditionTypes []
+
+	}
+	if len(slowStageMessages) > 0 {
+		output := fmt.Sprintf("%d nodes took over 5 minutes to stage OSUpdate:\n\n%s",
```
might be a local variable for `5*time.Minute` that we can format in here, so folks don't bump the threshold but forget to bump this output string.

deads2k commented:
/lgtm

deads2k commented:
> @wking has pointed out that we may have suites that do multi-stage updates, meaning we move through these conditions several times. I'm also not sure this currently properly handles install time events. Will revisit asap tomorrow.

followup is acceptable to me.

deads2k commented:
cleaning up CI signal on tests

dgoodwin commented:
/hold 

Something's wrong

dgoodwin commented:
Fixed a pointer bug causing zero start times and empty staged times, added vars for the flake/fail thresholds, added the check for edge case on zero start time (which will fail the test for now), formatted timestamps better for printing, and *may* have handled test suites with multiple update cycles.

deads2k commented:
/lgtm

dgoodwin commented:
/hold cancel
